L1Trigger/Phase2L1ParticleFlow/plugins/L1TCtL2EgProducer.cc

```diff
@@ -352,27 +371,41 @@ void L1TCtL2EgProducer::convertToEmu(const l1t::TkEm &tkem,
   emu.sta_idx = refRemapper.origRefAndPtr.size() - 1;
-  emu.setHwIso(EGIsoObjEmu::IsoType::TkIso, l1ct::Scales::makeIso(tkem.trkIsol()));
-  emu.setHwIso(EGIsoObjEmu::IsoType::PfIso, l1ct::Scales::makeIso(tkem.pfIsol()));
-  emu.setHwIso(EGIsoObjEmu::IsoType::TkIsoPV, l1ct::Scales::makeIso(tkem.trkIsolPV()));
-  emu.setHwIso(EGIsoObjEmu::IsoType::PfIsoPV, l1ct::Scales::makeIso(tkem.pfIsolPV()));
+  // NOTE: The emulator and FW data-format stores absolute iso while the CMSSW object stores relative iso
+  emu.setHwIso(EGIsoObjEmu::IsoType::TkIso, l1ct::Scales::makeIso(tkem.trkIsol() * tkem.pt()));
+  emu.setHwIso(EGIsoObjEmu::IsoType::PfIso, l1ct::Scales::makeIso(tkem.pfIsol() * tkem.pt()));
+  emu.setHwIso(EGIsoObjEmu::IsoType::PuppiIso, l1ct::Scales::makeIso(tkem.puppiIsol() * tkem.pt()));
+  emu.setHwIso(EGIsoObjEmu::IsoType::TkIsoPV, l1ct::Scales::makeIso(tkem.trkIsolPV() * tkem.pt()));
+  emu.setHwIso(EGIsoObjEmu::IsoType::PfIsoPV, l1ct::Scales::makeIso(tkem.pfIsolPV() * tkem.pt()));
+  emu.setHwIso(EGIsoObjEmu::IsoType::PuppiIsoPV, l1ct::Scales::makeIso(tkem.puppiIsolPV() * tkem.pt()));
   // std::cout << "[convertToEmu] TkEM pt: " << emu.hwPt << " eta: " << emu.hwEta << " phi: " << emu.hwPhi << " staidx: " << emu.sta_idx << std::endl;
   boarOut.egphoton.push_back(emu);
 }
 
+void L1TCtL2EgProducer::convertToPuppi(const l1t::PFCandidateCollection &l1PFCands, l1ct::PuppiObjs &puppiObjs) const {
+  for (const auto &l1PFCand : l1PFCands) {
+    l1ct::PuppiObj obj;
+    obj.initFromBits(l1PFCand.encodedPuppi64());
+    puppiObjs.emplace_back(obj);
+  }
+}
+
 l1t::TkEm L1TCtL2EgProducer::convertFromEmu(const l1ct::EGIsoObjEmu &egiso, const RefRemapper &refRemapper) const {
   // std::cout << "[convertFromEmu] TkEm pt: " << egiso.hwPt << " eta: " << egiso.hwEta << " phi: " << egiso.hwPhi << " staidx: " << egiso.sta_idx << std::endl;
   // NOTE: the TkEM object is created with the accuracy as in GT object (not the Correlator internal one)!
   const auto gteg = egiso.toGT();
   reco::Candidate::PolarLorentzVector mom(
       l1gt::Scales::floatPt(gteg.v3.pt), l1gt::Scales::floatEta(gteg.v3.eta), l1gt::Scales::floatPhi(gteg.v3.phi), 0.);
+  // NOTE: The emulator and FW data-format stores absolute iso while the CMSSW object stores relative iso
   l1t::TkEm tkem(reco::Candidate::LorentzVector(mom),
                  refRemapper.origRefAndPtr[egiso.sta_idx].first,
                  egiso.floatRelIso(l1ct::EGIsoObjEmu::IsoType::TkIso),
                  egiso.floatRelIso(l1ct::EGIsoObjEmu::IsoType::TkIsoPV));
   tkem.setHwQual(gteg.quality);
   tkem.setPFIsol(egiso.floatRelIso(l1ct::EGIsoObjEmu::IsoType::PfIso));
   tkem.setPFIsolPV(egiso.floatRelIso(l1ct::EGIsoObjEmu::IsoType::PfIsoPV));
+  tkem.setPuppiIsol(egiso.floatRelIso(l1ct::EGIsoObjEmu::IsoType::PuppiIso));
+  tkem.setPuppiIsolPV(egiso.floatRelIso(l1ct::EGIsoObjEmu::IsoType::PuppiIsoPV));
```

Comment:
Probably not since it should be vertex constraint by construction already with PUPPI. I added them because the variable existed also in the TkEm data format (https://github.com/p2l1pfp/cmssw/blob/L1PF_12_3_X/DataFormats/L1TCorrelator/interface/TkEm.h#L77).
Shall we remove it here and there as well perhaps?

Reply:
My question was more about computing the PV-based isolation in the first place. We could maintain the functionality as optional (I would avoid changing the data-format for now), but I would avoid filling the variable by default since it is not going to happen in the current design. What do you think?